### PR TITLE
properly decode content-length

### PR DESCRIPTION
**Pull request by MartinNowak**
- split headers and body
- partly parse request header for content-length
  and handle 100-continue
- return headers and body separately
- fix missing content-length in patch method

### CODE REVIEW

File: std/net/curl.d [hunk 1]
```diff
@@ -237,17 +237,52 @@ version(unittest)
...
+                {
+                    nbytes = s.receive(tmp[0 .. min(remain, $)]);
+                    assert(nbytes >= 0);
+                    buf ~= tmp[0 .. nbytes];
+                    remain -= nbytes;
+                }
+            }
+            else
+            {
+                assert(bdy.empty);
```
Comment (JakobOvrum) [inline std/net/curl.d]: A failing `assert` is a bug in the source code. This looks like input validation.

File: std/net/curl.d [hunk 2]
```diff
@@ -237,17 +237,52 @@ version(unittest)
...
+            auto hdrs = cast(string)buf[0 .. $ - bdy.length];
+            bdy.popFrontN(4);
+            // no support for chunked transfer-encoding
+            if (auto m = hdrs.matchFirst(ctRegex!(`Content-Length: ([0-9]+)`, "i")))
+            {
+                import std.uni : asUpperCase;
+                if (hdrs.asUpperCase.canFind("EXPECT: 100-CONTINUE"))
+                    s.send(httpContinue);
+
+                size_t remain = m.captures[1].to!size_t - bdy.length;
```
Comment (JakobOvrum) [inline std/net/curl.d]: Maybe use the parsed content length to reduce the number of allocations for `buf`?

**Comment by JakobOvrum** [comment]
Nevermind my comments, I see now that this is test code.

**Comment by DmitryOlshansky** [comment]
LGTM

**Comment by MartinNowak** [comment]
Auto-merge toggled on

**Pull request merged**